File: frontend/src/api/compatibility.ts
import type { KeyboardDefinition, KeycapSet } from "../types/domain";
import type { CompatibilityResult } from "../types/compatibility";

export async function checkCompatibility(
  keyboard: KeyboardDefinition,
  keycapSet: KeycapSet,
  signal: AbortSignal,
): Promise<CompatibilityResult> {
  const response = await fetch("/api/compatibility/check", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ keyboard, keycapSet }),
    signal,
  });
  if (!response.ok) throw new Error(`Fit check failed (${response.status}).`);
  return response.json() as Promise<CompatibilityResult>;
}

File: backend/app/api/compatibility.py
from fastapi import APIRouter

from app.models.compatibility import CompatibilityRequest, CompatibilityResult
from app.services.compatibility import check_compatibility

router = APIRouter(tags=["compatibility"])


@router.post("/compatibility/check", response_model=CompatibilityResult)
def check(request: CompatibilityRequest) -> CompatibilityResult:
    return check_compatibility(request.keyboard, request.keycap_set)
